File: iosApp/iosApp/Features/Chat/ChatViewModel.swift
//
//  ChatViewModel.swift
//  BDE
//
//  Created by Nathan Fallet on 07/07/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class ChatViewModel: ObservableObject {
    
    @Published var conversations = [ChatConversation]()
    
    var onConversationChatMessage: ((ChatMessage) -> Void)?
    
    func onAppear(token: String?) {
        AnalyticsService.shared.log(.screenView(screenName: "chat", screenClass: "ChatView"))
        
        WebSocketService.shared.onWebSocketMessage = onWebSocketMessage
        fetchConversations(token: token)
    }
    
    func fetchConversations(token: String?) {
        guard let token else {
            return
        }
        Task {
            let conversations = try await CacheService.shared.apiService().getChat(token: token)
            DispatchQueue.main.async {
                self.conversations = conversations.sorted {
                    $0.lastMessage?.createdAt?.epochSeconds ?? 0 > $1.lastMessage?.createdAt?.epochSeconds ?? 0
                }
            }
        }
    }
    
    func onWebSocketMessage(message: Any) {
        DispatchQueue.main.async {
            switch (message) {
            case let chatMessage as ChatMessage:
                self.onChatMessage(chatMessage: chatMessage)
            case let chatMembership as ChatMembership:
                self.onChatMembership(chatMembership: chatMembership)
            default:
                break
            }
        }
    }
    
    func onChatMessage(chatMessage: ChatMessage) {
        conversations
            .filter { chatMessage.groupType == $0.groupType && chatMessage.groupId == $0.groupId }
            .forEach { $0.lastMessage = chatMessage }
        conversations = conversations.sorted {
            $0.lastMessage?.createdAt?.epochSeconds ?? 0 > $1.lastMessage?.createdAt?.epochSeconds ?? 0
        }
        onConversationChatMessage?(chatMessage)
    }
    
    func onChatMembership(chatMembership: ChatMembership) {
        conversations
            .filter { chatMembership.groupType == $0.groupType && chatMembership.groupId == $0.groupId }
            .forEach { $0.membership = chatMembership }
        conversations = conversations.sorted {
            $0.lastMessage?.createdAt?.epochSeconds ?? 0 > $1.lastMessage?.createdAt?.epochSeconds ?? 0
        }
    }
    
}

File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/chat/ChatViewModel.kt
package me.nathanfallet.bdeensisa.features.chat

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.launch
import me.nathanfallet.bdeensisa.database.DatabaseDriverFactory
import me.nathanfallet.bdeensisa.extensions.SharedCacheService
import me.nathanfallet.bdeensisa.models.ChatConversation
import me.nathanfallet.bdeensisa.models.ChatMembership
import me.nathanfallet.bdeensisa.models.ChatMessage
import me.nathanfallet.bdeensisa.services.WebSocketService

class ChatViewModel(
    application: Application,
    token: String?
) : AndroidViewModel(application) {

    // Properties

    private val conversations = MutableLiveData<List<ChatConversation>>()

    // Getters

    fun getConversations(): LiveData<List<ChatConversation>> {
        return conversations
    }

    // Methods

    init {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, "chat")
            param(FirebaseAnalytics.Param.SCREEN_CLASS, "ChatView")
        }

        WebSocketService.getInstance(getApplication()).onWebSocketMessage = this::onWebSocketMessage
        fetchConversations(token)
    }

    fun fetchConversations(token: String?) {
        if (token == null) {
            return
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getChat(token).let {
                        conversations.value = it.sortedByDescending { conversation ->
                            conversation.lastMessage?.createdAt?.epochSeconds ?: 0
                        }
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun onWebSocketMessage(message: Any) {
        viewModelScope.launch {
            when (message) {
                is ChatMessage -> onChatMessage(message)
                is ChatMembership -> onChatMembership(message)
            }
        }
    }

    fun onChatMessage(chatMessage: ChatMessage) {
        conversations.value
            ?.filter { chatMessage.groupType == it.groupType && chatMessage.groupId == it.groupId }
            ?.forEach { it.lastMessage = chatMessage }
        conversations.value = conversations.value?.sortedByDescending { conversation ->
            conversation.lastMessage?.createdAt?.epochSeconds ?: 0
        }
    }

    fun onChatMembership(chatMembership: ChatMembership) {
        conversations.value
            ?.filter { chatMembership.groupType == it.groupType && chatMembership.groupId == it.groupId }
            ?.forEach { it.membership = chatMembership }
        conversations.value = conversations.value?.sortedByDescending { conversation ->
            conversation.lastMessage?.createdAt?.epochSeconds ?: 0
        }
    }

}